### ISSUE TITLE
Links are not clearly identifiable - Body text/links (Accessibility)

### ISSUE THREAD

**Issue by dbranchini** [sul-dlss/SearchWorks#3779]
Color alone is not enough to distinguish a link from the surrounding text.

[SiteImprove Issue with details](https://my2.siteimprove.com/Accessibility/1059585/NextGen/Issue/1?ruleId=62&pageSegments=&issueKind=1&exceptTags=1,2&conformance=&siteTargetIssueKinds=1,2&wcagVersion=21&siteTarget.wcagVersion=21&siteTarget.conformanceLevels=0,1,3,4&siteTarget.issueKinds=1,2&siteTarget.wcagVersion=21&siteTarget.conformanceLevels=0,1,3,4&siteTarget.issueKinds=1,2
)

Note that on [this page](https://searchworks.stanford.edu/), it's only calling out two links even though there are several links on this page. Headings seem to be excluded, but what about the facet links? 

**To-do's**
- [x] @dbranchini: Clarify how to address. 

**Solution**
Adding an underline is a possible solution.

**Comment by jcoyne** [comment]
@dbranchini please note that these links do have an underline in the hover state, so if you want them underlined by default, what should the hover state be?

**Comment by dbranchini** [comment]
Note this [WebAIM article](https://webaim.org/articles/contrast/#only) which states:

For optimal accessibility and usability, maintain the underline on links.  **Otherwise, two conditions must be met:**

- 3:1 contrast between the body text and the link text.
- A "visual cue" (not just a color change) that appears on mouse hover and keyboard focus. The most common way to meet this is to underline the link on hover and focus.

The second criterion is met; however, the first one is not. [This tool is useful to test contrast](https://webaim.org/resources/linkcontrastchecker/) against the background color **AND between the body text color and the link text color**. 

Currently, we have the following:
- Background is `#FFFFFF` 
- Body text is `#212529` 
- Link text is `#0068B8`

The link text to background is fine; however, the link to body text is not. It's 2.69:1. **If we change the body text to `#000000`, then the ratio is 3.67:1, which passes and this is my proposed solution for any trailing links and/or links within a body of text.** In reading through the WebAIM documentation, these requirements are specifically for links within a body of text; therefore, the facets, heading and/or navigation links are not affected.

**Comment by jcoyne** [comment]
> In reading through the WebAIM documentation, these requirements are specifically for links within a body of text; therefore, the facets, heading and/or navigation links are not affected.

So, we don't need to do anything?

**Comment by dbranchini** [comment]
Yes, that's true. I also noticed we use #212529 for a lot of body text. It seems most of our links are not embedded in body text (as you say above in this case), so we could probably ignore this issue raised by SiteImprove. I originally thought changing the body text from #212529 to #000 would be a win overall (it helps with contrast against the background color on the record view), but let's skip it and close it for now. There's talk of a facelift at a later time, and we can consider this at that time since the text against the background isn't being flagged at the AA level and this is a false positive if we consider it's not embedded in body text.

**Comment by dbranchini** [comment]
After the commentary above, I just realized that it's flagged on other pages too where it is referencing a link embedded in body text:

https://searchworks.stanford.edu/articles (note the link "log in now" in the off-campus section and https://searchworks.stanford.edu/selected_databases (not the link "Learn more about finding articles" in the Selected article database section at the top. 

If we change all body text to a standard black #000, it seems like a win overall for contrast.

**Comment by jcoyne** [comment]
@dbranchini I believe Gary selected that color to reduce eyestrain against the background. See https://uxplanet.org/basicdesign-never-use-pure-black-in-typography-36138a3327a6. Could we instead style the links that are in body text with a standard underline decoration?

**Comment by dbranchini** [comment]
I understand your point, and at the same time, there's conflicting information. For example, in the article you reference above, there's this statement:

> Pure black isn’t harmful to all users. Low-vision users, who are sight impaired but not blind, tend to read text better with pure black text or white text on a black background. When designing for them, you may need to use black for an accessibility mode. 

And there's this article, https://userway.org/blog/digital-eye-strain/, which states the following solutions for eye strain: larger fonts, sufficient line height spacing, **apply strong color contrast to your text and background**, use subheadings (for easier scanning), break out page with bullets and images, and structure content in short sentences and paragraphs. There's no mention of contrast. 

I've also thought about dark mode being a better solution to reduce eye strain - typically the #1 reason users switch to dark mode per this article, https://www.nngroup.com/articles/dark-mode-users-issues/#:~:text=Color%20Contrast,3%3A1%20for%20large%20text.

Anyway, this is a very long way of saying that I'm not sure there's a single solution to match all users' needs. The article you reference does point to research, whereas the first article I reference does not, so what about a compromise? I played around with the contrast checker, leaving the link color and background color the same, and then started with our body text/font color and then made it minimally darker to achieve the 3:1 ratio for link-to-body text colors, which resulted in `#1C2021` - see screenshot below. It wouldn't be pure black and we'd meet accessibility guidelines, although I don't know if SiteImprove is "smart" enough to clear the issue based on this change, so if you feel strongly about not touching it, then we'll leave it as is, and then address these thoughts in the future when/if a fuller visual update is made to SearchWorks.

![Image](https://github.com/sul-dlss/SearchWorks/assets/6816624/7d3692d7-18e1-4989-ae9a-a4537692d37b)

**Comment by jcoyne** [comment]
>  I'm not sure there's a single solution to match all users' needs. 

What about using `text-decoration: underline` to identify links in the body?  Then we don't need to make any changes regarding contrasts?

**Comment by dbranchini** [comment]
This solution sounds good to me, but what will it include exactly? For example, would it include the links on an item description in the record view - see attached screenshot. 

![Screenshot 2024-01-29 at 10 56 24 AM](https://github.com/sul-dlss/SearchWorks/assets/6816624/07434f87-9a70-424b-8a5e-431b61e9cdc1)

**Comment by jcoyne** [comment]
@dbranchini It certainly could. Is that desired?

**Comment by dbranchini** [comment]
It doesn't need to include those. What if it was constrained to links in `<p>` elements so that it's more likely to capture links surrounded by body text?

**Comment by dbranchini** [comment]
After discussion with @alundgard, @saseestone and @cbeer, we decided the best solution is to change the body text color to `#030303`. The reason it's even darker than the suggestion listed above is due to the cream background `#FBFBFC` on the record detail view. This allows it to pass against that background as well as a white background.

**Comment by jcoyne** [comment]
@dbranchini @alundgard You are aware that the #FBFBFC is from the obsolete brand palette,  which has since been updated to use white as the background color, right?  https://identity.stanford.edu/design-elements/color/.  We should be changing that some time.

**Comment by cbeer** [comment]
I'm confused where the current text color came from. Back in August, it was `#000`.

**Comment by jcoyne** [comment]
@cbeer it was from the bootstrap upgrade.  I had thought of this as an improvement that we got for free by upgrading rather than a regression.  
https://github.com/twbs/bootstrap/blob/v4.6.2/scss/_reboot.scss#L52
https://github.com/twbs/bootstrap/blob/v4.6.2/scss/_variables.scss#L169
https://github.com/twbs/bootstrap/blob/v4.6.2/scss/_variables.scss#L17

**Comment by dbranchini** [comment]
@jcoyne , I am aware of that - unfortunately, a lot of the site is not using brand guidelines. I also tried Stanford's black and grays with their digital blue against a white background, and it doesn't pass for this particular issue. It seems we need a fuller review of the color palette. I'm happy to do this if @saseestone wants me to. However, changing the text color to #000 (black) does fix it.

**Comment by jcoyne** [comment]
@dbranchini Why have we eliminated underline under links in the body text as an option?  This seems like the ideal way as it is not dependent on contrast at all. This allows us to move to the colors defined in the brand standard and fix the accessability bug.

**Comment by dbranchini** [comment]
In my opinion, when that solution was tried, underlines were applied broadly, and it created visual clutter. As you mention, even our background color is not within the Stanford guidelines. I feel black text against this cream background works until we look at a fuller redesign of the site.

**Comment by dbranchini** [comment]
After my discussion, we found a solution everyone seems more comfortable with. We will change the primary font color to Stanford Black `#2E2D29`, the links will remain Digital Blue `#006CB8`, and the underline will match the PURL redesign `#adb5bd`. This will include everything, except headers that are links, such as the item title on the record view. See screenshots below:


![Image](https://github.com/sul-dlss/SearchWorks/assets/6816624/e9f0cd9c-7d3d-4a64-acc8-0422aac9a92f)
![Image](https://github.com/sul-dlss/SearchWorks/assets/6816624/dbd73b0c-7527-4839-8234-49e4e74aa2b2)

**Comment by dbranchini** [comment]
Thanks for asking this clarifying question, @cbeer, and yes, you are correct. No need to add an underline in the top navigation or footer either.
